File: BattleshipCS/GameManager.cs
namespace BattleshipCS;
public class GameManager
{
    private readonly Player player1;
    private readonly Player player2;
    private Player currentPlayer;
    private bool gameOver;
    private readonly UserInterface userInterface;

    public GameManager(int boardSize)
    {
        gameOver = false;
        userInterface = new UserInterface(this);

        player1 = new HumanPlayer("Игрок 1", boardSize);
        player2 = new AIPlayer("Компьютер", boardSize);

        currentPlayer = player1;

        // Устанавливаем ссылки на поля противников
        player1.SetEnemyBoard(player2.MyBoard);
        player2.SetEnemyBoard(player1.MyBoard);
    }

    public void SetupGame()
    {
        Console.WriteLine($"Расстановка кораблей для {player1.Name}:");
        player1.PlaceShips();

        Console.WriteLine($"Расстановка кораблей для {player2.Name}:");
        player2.PlaceShips();

        Console.WriteLine("Игра начинается!");
    }

    public void RunGameLoop()
    {
        while (!gameOver)
        {
            var aiPlayer = currentPlayer as AIPlayer;

            // Показываем состояние после хода
            if (aiPlayer == null)
            {
                DisplayGameState();
            }

            // Ход текущего игрока
            var move = currentPlayer.MakeMove();

            // Обработка выстрела
            var enemyBoard = currentPlayer.EnemyBoard;
            var result = enemyBoard!.ReceiveShot(move);

            // Обновление состояния ИИ если нужно
            if (aiPlayer != null)
            {
                aiPlayer.UpdateAIState(result, move);
            }

            // Отображение результата
            Console.WriteLine($"{currentPlayer.Name} стреляет в ({move.Item1}, {move.Item2}) - {GetResultMessage(result)}");

            // Показываем состояние после хода
            if (aiPlayer == null)
            {
                DisplayGameState();
            }

            // Проверка окончания игры
            if (enemyBoard.IsAllShipsSunk())
            {
                gameOver = true;
                Console.WriteLine("\n=== ИГРА ОКОНЧЕНА ===");
                Console.WriteLine($"{currentPlayer.Name} ПОБЕДИЛ!");
                userInterface.ShowGameOver(currentPlayer.Name);
                break;
            }

            // Смена хода если не попадание
            if (result != Ship.ShotResult.Hit && result != Ship.ShotResult.Sunk)
            {
                SwitchTurn();
            }

            // Пауза для удобства восприятия
            if (aiPlayer == null)
            {
                Console.WriteLine("Нажмите Enter для продолжения...");
                Console.ReadLine();
            }
        }
    }

    private string GetResultMessage(Ship.ShotResult result)
    {
        return result switch
        {
            Ship.ShotResult.Hit => "ПОПАДАНИЕ!",
            Ship.ShotResult.Sunk => "КОРАБЛЬ ПОТОПЛЕН!",
            Ship.ShotResult.Miss => "ПРОМАХ!",
            Ship.ShotResult.AlreadyShot => "Уже стреляли сюда!",
            _ => "Неизвестный результат"
        };
    }

    private void SwitchTurn()
    {
        currentPlayer = currentPlayer == player1 ? player2 : player1;
    }

    private void DisplayGameState()
    {
        Console.WriteLine();
        userInterface.DisplayBoards(currentPlayer);
    }

    public Player CurrentPlayer => currentPlayer;
    public Player Player1 => player1;
    public Player Player2 => player2;
}

File: Battleship/GameManager.cpp
// GameManager.cpp
#include "GameManager.hpp"
#include "UserInterface.hpp"
#include <iostream>
#include <sstream>

GameManager::GameManager(int boardSize)
    : m_gameOver(false)
    , m_userInterface(std::make_unique<UserInterface>(this))
{
    // Создаем игроков через умные указатели
    m_player1 = std::make_unique<HumanPlayer>("Игрок 1", boardSize);
    m_player2 = std::make_unique<AIPlayer>("Компьютер", boardSize);

    // Текущий игрок - просто указатель (не владеющий)
    m_currentPlayer = m_player1.get();

    // Устанавливаем ссылки на поля противников через shared_ptr
    m_player1->SetEnemyBoard(std::make_shared<GameBoard>(m_player2->GetMyBoard()));
    m_player2->SetEnemyBoard(std::make_shared<GameBoard>(m_player1->GetMyBoard()));
}

void GameManager::SetupGame()
{
    std::cout << "Расстановка кораблей для " << m_player1->GetName() << ":\n";
    m_player1->PlaceShips();

    std::cout << "Расстановка кораблей для " << m_player2->GetName() << ":\n";
    m_player2->PlaceShips();

    std::cout << "Игра начинается!\n";
}

void GameManager::RunGameLoop()
{
    while (!m_gameOver)
    {
        // Используем dynamic_cast с сырыми указателями
        AIPlayer* aiPlayer = dynamic_cast<AIPlayer*>(m_currentPlayer);

        // Показываем состояние после хода
        if (!aiPlayer)
        {
            DisplayGameState();
        }

        // Ход текущего игрока
        Player::MoveType move = m_currentPlayer->MakeMove();

        // Обработка выстрела - получаем shared_ptr и разыменовываем
        std::shared_ptr<GameBoard> enemyBoard = m_currentPlayer->GetEnemyBoard();
        Ship::ShotResult result = enemyBoard->ReceiveShot(move);

        // Обновление состояния ИИ если нужно
        if (aiPlayer)
        {
            aiPlayer->UpdateAIState(result, move);
        }

        // Отображение результата
        std::cout << m_currentPlayer->GetName() << " стреляет в ("
            << move.first << ", " << move.second << ") - ";

        switch (result)
        {
        case Ship::ShotResult::eHit:
            std::cout << "ПОПАДАНИЕ!\n";
            break;
        case Ship::ShotResult::eSunk:
            std::cout << "КОРАБЛЬ ПОТОПЛЕН!\n";
            break;
        case Ship::ShotResult::eMiss:
            std::cout << "ПРОМАХ!\n";
            break;
        case Ship::ShotResult::eAlreadyShot:
            std::cout << "Уже стреляли сюда!\n";
            break;
        }

        // Показываем состояние после хода
        if (!aiPlayer)
        {
            DisplayGameState();
        }

        // Проверка окончания игры
        if (enemyBoard->IsAllShipsSunk())
        {
            m_gameOver = true;
            std::cout << "\n=== ИГРА ОКОНЧЕНА ===\n";
            std::cout << m_currentPlayer->GetName() << " ПОБЕДИЛ!\n";
            m_userInterface->ShowGameOver(m_currentPlayer->GetName());
            break;
        }

        // Смена хода если не попадание
        if (result != Ship::ShotResult::eHit && result != Ship::ShotResult::eSunk)
        {
            SwitchTurn();
        }

        // Пауза для удобства восприятия
        if (!aiPlayer)
        {
            std::cout << "Нажмите Enter для продолжения...";
            std::cin.ignore();
            std::cin.get();
        }
    }
}

void GameManager::SwitchTurn()
{
    if (m_currentPlayer == m_player1.get())
    {
        m_currentPlayer = m_player2.get();
    }
    else
    {
        m_currentPlayer = m_player1.get();
    }
}

void GameManager::DisplayGameState()
{
    std::cout << "\n";
    m_userInterface->DisplayBoards(m_currentPlayer);
}